### PR TITLE
Add check for existing Argo Workflows

### PR DESCRIPTION
**Pull request by VaishnaviHire**
## Description
Jira Issue: https://issues.redhat.com/browse/RHOAIENG-4570
This PR introduces changes to handle existing Argo workflows for new install and upgrade. Following is the expected workflow 

```
For version n

1. New Install
    - Argo Workflow exists
        - All components in n deployed successfully except DSP
        - New Dashboard(n) shows warning banner

2. Upgrade from n-1 --> n
   - Argo workflow exists and DSP is enabled
      - Do not upgrade any component. All components in (n-1)
      - Dashboard will not show any warning
      - DSC(Operator) will trigger events stating remove Argo or disable DSP to upgrade to new version
      - Note, (n-1) manifests are no longer managed here

``` 
### New Install

1. Deploy [Workflow CRD ](https://github.com/opendatahub-io/data-science-pipelines-operator/blob/v2.0.1/config/argo/crd.workflows.yaml).
2. Deploy operator using catalogsource: quay.io/vhire/opendatahub-operator-catalog:v2.10.1
3. Create DSCI and DSC with DataSciecnePipelines enabled
4. Verify all components except DSP are updated
5. Verify conditiontype `CapabilityDSPv2Argo`  is triggered .
6. Verify DSP status is False

### Upgrade

1.  Install an ODH version with v1 dsp pipelines(e.g uay.io/vhire/opendatahub-operator-catalog:v2.7.0-711)
2.  Deploy [Workflow CRD ](https://github.com/opendatahub-io/data-science-pipelines-operator/blob/v2.0.1/config/argo/crd.workflows.yaml)
3. Create DSCi and DSC with DataSciecnePipelines enabled
4. Replace catalog source image with new version quay.io/vhire/opendatahub-operator-catalog:v2.10.1
5. Verify no components are upgraded to new version
6. Verify conditiontype `CapabilityDSPv2Argo`  is triggered .
8. Verify conditiontype `Upgradeable` is set to `False`

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: components/datasciencepipelines/datasciencepipelines.go [hunk 1]
```diff
@@ -141,3 +149,20 @@ func (d *DataSciencePipelines) ReconcileComponent(ctx context.Context,
 
 	return nil
 }
+
+func argoWorkflowDeploymentAllowed(ctx context.Context,
```
Comment (zdtsw) [inline components/datasciencepipelines/datasciencepipelines.go:153]: should we rename this function to
`unmanagedArgoWFExist`

File: components/datasciencepipelines/datasciencepipelines.go [hunk 2]
```diff
@@ -106,6 +109,11 @@ func (d *DataSciencePipelines) ReconcileComponent(ctx context.Context,
 		}
 	}
 
+	// Check for existing Argo Workflows
```
Comment (zdtsw) [inline components/datasciencepipelines/datasciencepipelines.go:112]: should this check be moved a bit up? like in line 96 before` if enabled`

File: components/datasciencepipelines/datasciencepipelines.go [hunk 3]
```diff
@@ -141,3 +149,20 @@ func (d *DataSciencePipelines) ReconcileComponent(ctx context.Context,
...
+			return nil
+		}
+		return fmt.Errorf("failed to get Workflow CRD : %w", err)
+	}
+	// Verify if existing workflow is deployed by ODH
+	_, odhLabelExists := workflowCRD.Labels[labels.ODH.Component(ComponentName)]
+	if odhLabelExists {
+		return nil
+	}
+	return fmt.Errorf("failed to deploy DSP. Workflow CRD already exists. Remove CRD to upgrade")
```
Comment (zdtsw) [inline components/datasciencepipelines/datasciencepipelines.go:167]:
```suggestion
	return fmt.Errorf("failed to deploy DSP. Argo Workflow CRD already exists but not deployed by our operator . Remove CRD to upgrade")
```

**Review by zdtsw — COMMENTED** [review]
some nitty gritty but i am fine with the change.

one more thing, what will happen from Dashboard side.
if we keep throwing error of  "CRD already there, go remove it" will user still be able to see the feature of enable DSPV2 from dashboard?

File: components/datasciencepipelines/datasciencepipelines.go (hunk 3 above)
Comment (etirelli) [reply components/datasciencepipelines/datasciencepipelines.go:167]: nitpick: replace "our" by "this" or "ODH".

**Comment by VaishnaviHire** [comment]
> some nitty gritty but i am fine with the change.
> 
> one more thing, what will happen from Dashboard side. if we keep throwing error of "CRD already there, go remove it" will user still be able to see the feature of enable DSPV2 from dashboard?

We still will have v1 manifests deployed. @lucferbux Do you know what will be the dashboard behavior in this case?

**Comment by kywalker-rh** [comment]
Can you provide a screenshot of where/when this will be shown? Also, it would be good to provide the user with a link to further documentation about why exactly they need to remove the CRD/operator.

**Review by zdtsw — APPROVED** [review]

**Comment by zdtsw** [comment]
/onhold
just to get more review comments, feel free to remove label @VaishnaviHire

**Comment by VaishnaviHire** [comment]
> Can you provide a screenshot of where/when this will be shown? Also, it would be good to provide the user with a link to further documentation about why exactly they need to remove the CRD/operator.

This is shown under `Details` for DataScienceCluster resource. I have attached a screen grab
[CheckArgoWorkflows.mov.zip](https://github.com/opendatahub-io/opendatahub-operator/files/14887511/CheckArgoWorkflows.mov.zip)

To resolve this error users will need to delete the Workflow crd by running -
```
oc delete crd workflows.argoproj.io
```

**Comment by kywalker-rh** [comment]
Thanks for sharing that @VaishnaviHire, I'm worried that's pretty buried though. Will an admin typically check those statuses after starting an upgrade?

With the status set as Ready I wouldn't expect to need to dig into it further to find out something failed. I want to prevent users being frustrated and confused as to why Pipelines won't work. Would it make sense to add a Banner for admins to see in the dashboard?

**Comment by zdtsw** [comment]
> Thanks for sharing that @VaishnaviHire, I'm worried that's pretty buried though. Will an admin typically check those statuses after starting an upgrade?
> 
> With the status set as Ready I wouldn't expect to need to dig into it further to find out something failed. I want to prevent users being frustrated and confused as to why Pipelines won't work. Would it make sense to add a Banner for admins to see in the dashboard?

should we move this discussion into original JIRA to get more eyes on?
changes on Dashboard side, will need UI team's work

File: controllers/datasciencecluster/datasciencecluster_controller.go [hunk 4]
```diff
@@ -281,12 +296,15 @@ func (r *DataScienceClusterReconciler) reconcileSubComponent(ctx context.Context
 		instance = r.reportError(err, instance, "failed to reconcile "+componentName+" on DataScienceCluster")
 		instance, _ = status.UpdateWithRetry(ctx, r.Client, instance, func(saved *dsc.DataScienceCluster) {
 			if enabled {
-				status.SetComponentCondition(&saved.Status.Conditions, componentName, status.ReconcileFailed, fmt.Sprintf("Component reconciliation failed: %v", err), corev1.ConditionFalse)
+				if componentName == datasciencepipelines.ComponentName && strings.Contains(err.Error(), datasciencepipelines.ArgoWorkflowCRD) {
```
Comment (zdtsw) [inline controllers/datasciencecluster/datasciencecluster_controller.go:299]: we probably no need check componentname here, 
```suggestion
				if strings.Contains(err.Error(), datasciencepipelines.ArgoWorkflowCRD) {
```

**Review by asanzgom — APPROVED** [review]

**Comment by Gkrumbach07** [comment]
This condition looks good from my side where we are testing against:
```tsx
condition.type === 'CapabilityDSPv2Argo' && condition.status === 'False',
```

**Comment by zdtsw** [comment]
tested again with latest code in the negative case
Screenshot from 2024-04-12 16-55-48

**Review by asanzgom — APPROVED** [review]

**Comment by VaishnaviHire** [comment]
@etirelli Added the error phase
Screenshot 2024-04-12 at 2 43 32 PM

File: controllers/datasciencecluster/datasciencecluster_controller.go [hunk 5]
```diff
@@ -446,3 +468,16 @@ func (r *DataScienceClusterReconciler) watchDataScienceClusterResources(a client
 
 	return nil
 }
+
+// argoWorkflowCRDPredicates filters the delete events to trigger reconcile when Argo Workflow CRD is deleted.
+var argoWorkflowCRDPredicates = predicate.Funcs{
+	DeleteFunc: func(e event.DeleteEvent) bool {
+		if e.Object.GetName() == datasciencepipelines.ArgoWorkflowCRD {
+			labels := e.Object.GetLabels()
+			if _, ok := labels["app.opendatahub.io/"+datasciencepipelines.ComponentName]; !ok {
```
Comment (bartoszmajsak) [inline controllers/datasciencecluster/datasciencecluster_controller.go:477]: We could use `labels.ODH.Component(datasciencepipelines.ComponentName)`

Comment (zdtsw) [reply controllers/datasciencecluster/datasciencecluster_controller.go:477]: i have it updated in https://github.com/opendatahub-io/opendatahub-operator/pull/967/files#diff-ddd4fc79031732c6d6cbc28c4f3fae64ee035ff75633e6d19097f717f36311b6

**Pull request merged**